src/renderer/viz/expressions/RampGeneric.js

```diff
@@ -0,0 +1,168 @@
+import { checkType, checkInstance, mix, fract } from './utils';
+
+import NamedColor from './color/NamedColor';
+import Property from './basic/property';
+import ListImage from './ListImage';
+import Linear from './linear';
+import CIELabGLSL from './color/CIELab.glsl';
+import CategoryIndex from './CategoryIndex';
+import { constant } from '../expressions';
+import { OTHERS_GLSL_VALUE, OTHERS_INDEX, DEFAULT_OPTIONS, DEFAULT_OTHERS } from './constants';
+import Palette from './color/palettes/Palette';
+import Base from './base';
+
+const paletteTypes = {
+    PALETTE: 'palette',
+    COLOR_ARRAY: 'color-list',
+    NUMBER_ARRAY: 'number-list',
+    IMAGE_LIST: 'image-list'
+};
+
+const inputTypes = {
+    NUMBER: 'number',
+    CATEGORY: 'category'
+};
+
+export default class RampGeneric extends Base {
+    _bindMetadata (metadata) {
+        super._bindMetadata(metadata);
+
+        this.type = this.palette.childType;
+        if (this.others === DEFAULT_OTHERS) {
+            this.others = this.palette.type === 'number-list' ? constant(1) : new NamedColor('gray');
+        } else {
+            checkType('ramp', 'others', 2, this.palette.childType, this.others);
+        }
+        if (this.input.isA(Property)) {
+            this.input = this.input.type === inputTypes.NUMBER
+                ? new Linear(this.input)
+                : new CategoryIndex(this.input);
+
+            this.input._bindMetadata(metadata);
+        }
+
+        checkType('ramp', 'input', 0, Object.values(inputTypes), this.input);
+
+        if (this.palette.type === paletteTypes.IMAGE_LIST) {
+            checkType('ramp', 'input', 0, inputTypes.CATEGORY, this.input);
+            checkInstance('ramp', 'palette', 1, ListImage, this.palette);
+        }
+
+        this.others._bindMetadata(metadata);
+        this.childrenNames.push('others');
+
+        this._properties = metadata.properties;
+        this._texCategories = null;
+        this._GLtexCategories = null;
+        this._metadata = metadata;
+    }
+
+    eval (feature) {
+        const input = this.input.eval(feature);
+        return this._calcEval(input, feature);
+    }
+
+    _calcEval (input, feature) {
+        const maxPaletteSize = this.input.numCategoriesWithoutOthers < this.palette.length
```

Comment:
I don't remember why this was needed

Reply:
It is calculated to discard unneeded colors in color arrays, and it is only being used when the palette is a color array: `this.palette.eval(feature).slice(0, maxPaletteSize);`